File: python_tests/test_i2c_proxy.py
# -*- coding: utf-8 -*-
import pytest
from conftest import make_resp
from rust_shdlc_driver import (
    AsyncShdlcConnection,
    ShdlcConnection,
    ShdlcMockPort,
)
from sensirion_uart_sensorbridge import (
    AsyncSensorBridgeI2cProxy,
    AsyncSensorBridgeShdlcDevice,
    SensorBridgeI2cProxy,
    SensorBridgePort,
    SensorBridgeShdlcDevice,
)


def test_sync_i2c_proxy():
    port = ShdlcMockPort(bitrate=460800)
    conn = ShdlcConnection(port)
    bridge = SensorBridgeShdlcDevice(conn, slave_address=0)
    proxy = SensorBridgeI2cProxy(bridge, SensorBridgePort.ONE)

    assert proxy.description == "SensorBridge"
    assert proxy.channel_count is None

    # Success transceive
    port.push_rx_data(make_resp(0, 0x11, 0x00, b"\xaa\xbb"))
    status, err, rx = proxy.transceive(
        slave_address=0x70,
        tx_data=b"\x01",
        rx_length=2,
        read_delay=0.001,
        timeout=0.05,
    )
    assert status == SensorBridgeI2cProxy.STATUS_OK
    assert err is None
    assert rx == b"\xaa\xbb"

    # NACK error (0x29)
    port.push_rx_data(make_resp(0, 0x11, 0x29, b""))
    status, err, rx = proxy.transceive(
        slave_address=0x70,
        tx_data=b"\x01",
        rx_length=2,
        read_delay=0.001,
        timeout=0.05,
    )
    assert status == SensorBridgeI2cProxy.STATUS_NACK
    assert rx == b""

    # Timeout error (0x2B)
    port.push_rx_data(make_resp(0, 0x11, 0x2B, b""))
    status, err, rx = proxy.transceive(
        slave_address=0x70,
        tx_data=b"\x01",
        rx_length=2,
        read_delay=0.001,
        timeout=0.05,
    )
    assert status == SensorBridgeI2cProxy.STATUS_TIMEOUT
    assert rx == b""


@pytest.mark.asyncio
async def test_async_i2c_proxy():
    port = ShdlcMockPort(bitrate=460800)
    conn = AsyncShdlcConnection(port)
    bridge = AsyncSensorBridgeShdlcDevice(conn, slave_address=0)
    proxy = AsyncSensorBridgeI2cProxy(bridge, SensorBridgePort.TWO)

    assert proxy.description == "SensorBridge"
    assert proxy.channel_count is None

    # Success transceive
    port.push_rx_data(make_resp(0, 0x11, 0x00, b"\x11\x22"))
    status, err, rx = await proxy.transceive(
        slave_address=0x44,
        tx_data=b"\x00",
        rx_length=2,
        read_delay=0.001,
        timeout=0.05,
    )
    assert status == AsyncSensorBridgeI2cProxy.STATUS_OK
    assert err is None
    assert rx == b"\x11\x22"

File: tests/test_i2c_proxy.rs
use rust_shdlc_driver::connection::{AsyncShdlcConnection, ShdlcConnection};
use rust_shdlc_driver::protocol::*;
use rust_shdlc_driver::transport::MockTransport;
use sensirion_uart_sensorbridge::device::{
    AsyncSensorBridgeDevice, AsyncSensorBridgeI2cProxy, SensorBridgeDevice, SensorBridgeI2cProxy,
    I2C_STATUS_NACK, I2C_STATUS_OK,
};
use sensirion_uart_sensorbridge::protocol::definitions::SensorBridgePort;

fn make_response_frame(slave_addr: u8, cmd_id: u8, state: u8, data: &[u8]) -> Vec<u8> {
    let mut content = vec![slave_addr, cmd_id, state, data.len() as u8];
    content.extend_from_slice(data);
    let checksum = calculate_checksum(&content);
    content.push(checksum);
    let stuffed = stuff_data_bytes(&content);
    let mut raw = vec![START_STOP_BYTE];
    raw.extend_from_slice(&stuffed);
    raw.push(START_STOP_BYTE);
    raw
}

#[tokio::test]
async fn test_async_i2c_proxy() {
    let mock = MockTransport::new();
    let conn = AsyncShdlcConnection::new(Box::new(mock.clone()));
    let dev = AsyncSensorBridgeDevice::new(conn, 0x00);
    let mut proxy = AsyncSensorBridgeI2cProxy::new(dev, SensorBridgePort::One).unwrap();

    assert_eq!(proxy.description(), "SensorBridge");
    assert_eq!(proxy.channel_count(), None);
    assert_eq!(proxy.port(), SensorBridgePort::One);

    // Successful transceive
    mock.push_rx_data(&make_response_frame(0x00, 0x11, 0x00, &[0x12, 0x34]));
    let res = proxy.transceive(0x70, &[0xAA], 2, 0.001, 0.05).await;
    assert_eq!(res.status, I2C_STATUS_OK);
    assert_eq!(res.rx_data, vec![0x12, 0x34]);
    assert!(res.error.is_none());

    // NACK device error (error code 0x29)
    mock.push_rx_data(&make_response_frame(0x00, 0x11, 0x29, &[]));
    let res = proxy.transceive(0x70, &[0xAA], 2, 0.001, 0.05).await;
    assert_eq!(res.status, I2C_STATUS_NACK);
    assert!(res.rx_data.is_empty());
}

#[test]
fn test_sync_i2c_proxy() {
    let mock = MockTransport::new();
    let conn = ShdlcConnection::new(Box::new(mock.clone())).unwrap();
    let dev = SensorBridgeDevice::new(conn, 0x00);
    let mut proxy = SensorBridgeI2cProxy::new(dev, SensorBridgePort::Two).unwrap();

    mock.push_rx_data(&make_response_frame(0x00, 0x11, 0x00, &[0x55, 0x66]));
    let res = proxy.transceive(0x38, &[], 2, 0.0, 0.05);
    assert_eq!(res.status, I2C_STATUS_OK);
    assert_eq!(res.rx_data, vec![0x55, 0x66]);
}
